tests/drivers/input/kbd_matrix/src/main.c

```diff
@@ -0,0 +1,399 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/input/input.h>
+#include <zephyr/input/input_kbd_matrix.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/ztest.h>
+
+#define TEST_KBD_SCAN_NODE DT_INST(0, test_kbd_scan)
+
+/* test driver */
+
+static struct {
+	kbd_row_t rows[3];
+	int col;
+	bool detect_mode;
+} state;
+
+static void test_drive_column(const struct device *dev, int col)
+{
+	state.col = col;
+}
+
+static kbd_row_t test_read_row(const struct device *dev)
+{
+	return state.rows[state.col];
+}
+
+static void test_set_detect_mode(const struct device *dev, bool enabled)
+{
+	TC_PRINT("detect mode: enabled=%d\n", enabled);
+	state.detect_mode = enabled;
+}
+
+static const struct input_kbd_matrix_api test_api = {
+	.drive_column = test_drive_column,
+	.read_row = test_read_row,
+	.set_detect_mode = test_set_detect_mode,
+};
+
+INPUT_KBD_MATRIX_DT_DEFINE(TEST_KBD_SCAN_NODE);
+
+static const struct input_kbd_matrix_common_config
+	test_cfg = INPUT_KBD_MATRIX_DT_COMMON_CONFIG_INIT(
+			TEST_KBD_SCAN_NODE, &test_api);
+
+static struct input_kbd_matrix_common_data test_data;
+
+DEVICE_DT_DEFINE(TEST_KBD_SCAN_NODE, input_kbd_matrix_common_init, NULL,
+		 &test_data, &test_cfg,
+		 POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE, NULL);
+
+static const struct device *const test_dev = DEVICE_DT_GET(TEST_KBD_SCAN_NODE);
+
+/* support stuff */
+
+static const struct device *column_hook_last_dev;
+static int column_hook_last_col;
+
+void input_kbd_matrix_drive_column_hook(const struct device *dev, int col)
+{
+	column_hook_last_dev = dev;
+	column_hook_last_col = col;
+}
+
+static void state_set_rows(kbd_row_t r0, kbd_row_t r1, kbd_row_t r2)
```

Comment:
Because kbd_row_t is indexed by column, this is clearer as:
```suggestion
static void state_set_rows_by_column(kbd_row_t c0, kbd_row_t c1, kbd_row_t c2)
```

Reply:
Done.